include/fiction/io/print_layout.hpp

```diff
@@ -343,41 +353,147 @@ void print_charge_layout(std::ostream& os, const Lyt& lyt, const bool cs_color =
     static_assert(is_cell_level_layout_v<Lyt>, "Lyt is not a cell-level layout");
     static_assert(has_sidb_technology_v<Lyt>, "Lyt is not an SiDB layout");
     static_assert(has_siqad_coord_v<Lyt>, "Lyt is not based on SiQAD coordinates");
-    static_assert(has_get_charge_state_v<Lyt>, "Lyt does not implement the get_charge_state function");
+    static_assert(has_get_charge_state_v<Lyt> | has_get_sidb_defect_v<Lyt>,
+                  "Lyt does not implement the get_charge_state function");
 
     // empty layout
     if (lyt.is_empty())
     {
-        os << "[i] empty layout" << std::endl;
-        return;
+        if constexpr (has_get_sidb_defect_v<Lyt>)
+        {
+            if (lyt.num_defects() == 0)
+            {
+                os << "[i] empty layout" << std::endl;
+                return;
+            }
+        }
+        else
+        {
+            os << "[i] empty layout" << std::endl;
+            return;
+        }
     }
 
-    coordinate<Lyt> min{};
-    coordinate<Lyt> max{lyt.x(), lyt.y(), 1};
+    // determine the north-west and south-west cell in siqad-coordinates.
+    const auto [nw, se] = bounding_box_siqad<Lyt>(lyt);
 
-    if (crop_layout)
+    auto min_nw = nw;
+    auto max_se = se;
+
+    std::vector<typename Lyt::cell> defects{};
+
+    // Check if defects exist in the layout.
+    if constexpr (has_get_sidb_defect_v<Lyt>)
     {
-        const auto bb = bounding_box_2d{lyt};
 
+        defects.reserve(lyt.num_defects());
+        lyt.foreach_sidb_defect([&defects](const auto& c) { defects.push_back(c.first); });
+        std::sort(defects.begin(), defects.end());
+        if (!defects.empty())
```

Comment:
You could move the three statements inside the `if` branch

Reply:
Good point! Thanks!